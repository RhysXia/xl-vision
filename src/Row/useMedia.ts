import { useEffect, useState } from 'react'
import { isServer } from '../commons/utils/env'
import { voidFn } from '../commons/utils/function'

const matchMedia = (query: string) => {
  if (isServer) {
    return {
      addListener: voidFn,
      matches: false,
      removeListener: voidFn
    }
  }
  return window.matchMedia(query)
}

export type BreakPoint = 'xxl' | 'xl' | 'lg' | 'md' | 'sm' | 'xs'

// 顺序不能颠倒
export const breakPointMap: Record<BreakPoint, string> = {
  xxl: '(min-width: 1600px)',
  // tslint:disable-next-line
  xl: '(min-width: 1200px)',
  lg: '(min-width: 992px)',
  md: '(min-width: 768px)',
  sm: '(min-width: 576px)',
  xs: '(min-width: 0)'
}
export const breakPointArray: Array<BreakPoint> = Object.keys(breakPointMap) as Array<BreakPoint>

const useMedia = () => {
  const [media, setMedia] = useState<Partial<Record<BreakPoint, boolean>>>({})

  useEffect(() => {
      const handlerMap: {
        [breakPoint: string]: {
          listener: EventListener
          mediaQuery: ReturnType<typeof matchMedia>
        }
      } = {}
      for (const breakPoint of breakPointArray) {
        const query = breakPointMap[breakPoint]
        const mql = matchMedia(query)
        const onChange = () => {
          setMedia((prev) => ({
            ...prev,
            [breakPoint]: mql.matches
          }))
        }
        handlerMap[breakPoint] = {
          listener: onChange,
          mediaQuery: mql
        }
        mql.addListener(onChange)
        onChange()
      }
      return () => {
        for (const breakPoint of breakPointArray) {
          const { mediaQuery, listener } = handlerMap[breakPoint]
          mediaQuery.removeListener(listener)
        }
      }
    }, []
  )

  return media
}

export default useMedia
