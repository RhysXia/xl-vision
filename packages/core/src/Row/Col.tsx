import PropTypes from 'prop-types'
import React from 'react'
import clsx from 'clsx'
import { BreakPoint, breakPointArray } from './useMedia'
import RowContext from './RowContext'
import createUseStyles from '../styles/createUseStyles'
import { Styles } from '@xl-vision/styles'

export type ColSpanType = number | Partial<Record<BreakPoint, number>>

export type ClassesKey =
  'root'
  |
  'span-0'
  | 'span-1'
  | 'span-2'
  | 'span-3'
  | 'span-4'
  | 'span-5'
  | 'span-6'
  | 'span-7'
  | 'span-8'
  | 'span-9'
  | 'span-10'
  | 'span-11'
  | 'span-12'
  | 'span-13'
  | 'span-14'
  | 'span-15'
  | 'span-16'
  | 'span-17'
  | 'span-18'
  | 'span-19'
  | 'span-20'
  | 'span-21'
  | 'span-22'
  | 'span-23'
  | 'span-24'
  |
  'offset-1'
  | 'offset-2'
  | 'offset-3'
  | 'offset-4'
  | 'offset-5'
  | 'offset-6'
  | 'offset-7'
  | 'offset-8'
  | 'offset-9'
  | 'offset-10'
  | 'offset-11'
  | 'offset-12'
  | 'offset-13'
  | 'offset-14'
  | 'offset-15'
  | 'offset-16'
  | 'offset-17'
  | 'offset-18'
  | 'offset-19'
  | 'offset-20'
  | 'offset-21'
  | 'offset-22'
  | 'offset-23'
  | 'offset-24'
  |
  'push-1'
  | 'push-2'
  | 'push-3'
  | 'push-4'
  | 'push-5'
  | 'push-6'
  | 'push-7'
  | 'push-8'
  | 'push-9'
  | 'push-10'
  | 'push-11'
  | 'push-12'
  | 'push-13'
  | 'push-14'
  | 'push-15'
  | 'push-16'
  | 'push-17'
  | 'push-18'
  | 'push-19'
  | 'push-20'
  | 'push-21'
  | 'push-22'
  | 'push-23'
  | 'push-24'
  |
  'pull-1'
  | 'pull-2'
  | 'pull-3'
  | 'pull-4'
  | 'pull-5'
  | 'pull-6'
  | 'pull-7'
  | 'pull-8'
  | 'pull-9'
  | 'pull-10'
  | 'pull-11'
  | 'pull-12'
  | 'pull-13'
  | 'pull-14'
  | 'pull-15'
  | 'pull-16'
  | 'pull-17'
  | 'pull-18'
  | 'pull-19'
  | 'pull-20'
  | 'pull-21'
  | 'pull-22'
  | 'pull-23'
  | 'pull-24'
  |
  'order-1'
  | 'order-2'
  | 'order-3'
  | 'order-4'
  | 'order-5'
  | 'order-6'
  | 'order-7'
  | 'order-8'
  | 'order-9'
  | 'order-10'
  | 'order-11'
  | 'order-12'
  | 'order-13'
  | 'order-14'
  | 'order-15'
  | 'order-16'
  | 'order-17'
  | 'order-18'
  | 'order-19'
  | 'order-20'
  | 'order-21'
  | 'order-22'
  | 'order-23'
  | 'order-24'

export interface ColProps extends React.HTMLAttributes<HTMLDivElement> {
  children?: React.ReactNode
  className?: string
  offset?: ColSpanType
  order?: ColSpanType
  clsPrefix?: string
  pull?: ColSpanType
  push?: ColSpanType
  span: ColSpanType
  classes?: Partial<Record<ClassesKey, string>>
}

const Col: React.FunctionComponent<ColProps> = (props) => {
  const { media, gutter } = React.useContext(RowContext)

  const {
    className,
    children,
    style,
    span,
    order,
    offset,
    pull,
    push,
    classes,
    ...others
  } = props

  const builtinStyles = useStyles(classes)

  const rootClassName = React.useMemo(() => {
    const obj = {
      offset,
      order,
      pull,
      push,
      span
    }
    const arr = [builtinStyles.root]
    Object.keys(obj).forEach((prop) => {
      const propValue = obj[prop as keyof typeof obj]
      if (typeof propValue === 'number') {
        arr.push(builtinStyles[`${prop}-${propValue}`])
      } else if (typeof propValue === 'object') {
        breakPointArray.some((breakPoint) => {
          const value = propValue[breakPoint]
          if (media[breakPoint] && value !== undefined) {
            arr.push(builtinStyles[`${prop}-${value}`])
            // 只需要满足最高的条件
            return true
          }
          return false
        })
      }
    })
    return clsx(arr, className)
  }, [media, span, order, offset, pull, push, className, builtinStyles])

  const colStyle =
    gutter > 0
      ? {
        ...style,
        paddingLeft: gutter / 2,
        paddingRight: gutter / 2
      }
      : style

  return (
    <div role='cell' {...others} style={colStyle} className={rootClassName}>
      {children}
    </div>
  )
}

const spanValidator = (props: ColProps, propName: keyof ColProps, componentName: string) => {
  // eslint-disable-next-line react/destructuring-assignment
  const propValue = props[propName] as ColSpanType
  if (typeof propValue === 'undefined') {
    return null
  }
  if (typeof propValue === 'number') {
    if (propValue < 0 || propValue > 24) {
      return new Error(
        `prop '${propName}' supplied to '${componentName} should be in 0-24 but actually '${propValue}'.`
      )
    }
  } else if (typeof propValue === 'object') {
    for (let i = 0; i < breakPointArray.length; i++) {
      const breakPoint = breakPointArray[i]
      const val = propValue[breakPoint]
      if (!val) {
        return null
      }
      if (Number.isInteger(val)) {
        if (propValue < 0 || propValue > 24) {
          return new Error(
            `prop '${propName}' supplied to '${componentName}' is object, its prop '${breakPoint}' be in 0-24 but actually '${val}'.`
          )
        }
      } else {
        return new Error(
          `prop '${propName}' supplied to '${componentName}' is object, its prop '${breakPoint}' should be a number.`
        )
      }
    }
  } else {
    return new Error(
      // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
      `prop '${propName}' supplied to '${componentName}' should be a integer or suitable object. It's type can't be '${propValue}'`
    )
  }
  return null
}

Col.displayName = 'Col'

Col.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string,
  offset: spanValidator,
  order: spanValidator,
  clsPrefix: PropTypes.string,
  pull: spanValidator,
  push: spanValidator,
  span: spanValidator,
  style: PropTypes.object,
  classes: PropTypes.object,
}

export default Col


const useStyles = createUseStyles(theme => {
    const styles: Styles<string> = {
      root: {
        float: 'left',
        boxSizing: 'border-box'
      }
    }
    const { span } = theme
    for (let i = 0; i <= span; i++) {
      if (i === 0) {
        styles[`spin-${i}`] = {
          display: 'none'
        }
        continue
      }
      styles[`span-${i}`] = {
        display: 'block',
        width: `${i / span * 100}%`,
        minHeight: 1
      }
      styles[`offset-${i}`] = {
        marginLeft: `${i / span * 100}%`
      }
      styles[`push-${i}`] = {
        position: 'relative',
        left: `${i / span * 100}%`
      }
      styles[`pull-${i}`] = {
        position: 'relative',
        right: `${i / span * 100}%`
      }
      styles[`order-${i}`] = {
        order: i
      }
    }
    return styles
  },
  'Col'
)
