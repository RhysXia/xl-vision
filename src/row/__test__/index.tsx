import { render } from 'enzyme'
import * as React from 'react'
import Row from '..'

describe('grid', () => {
  it('基本用法', () => {
    const wrapper = render(
      <div>
        <div className='grid-row-box'>
          <Row gutter={10}>
            <Row.Col span={4}>
              <div className='grid-col-box'>col1</div>
            </Row.Col>
            <Row.Col span={5}>
              <div className='grid-col-box'>col2</div>
            </Row.Col>
            <Row.Col span={7}>
              <div className='grid-col-box'>col3</div>
            </Row.Col>
            <Row.Col span={8}>
              <div className='grid-col-box'>col4</div>
            </Row.Col>
          </Row>
        </div>
        <div className='grid-row-box'>
          <Row gutter={10}>
            <Row.Col span={6}>
              <div className='grid-col-box'>col1</div>
            </Row.Col>
            <Row.Col span={6}>
              <div className='grid-col-box'>col2</div>
            </Row.Col>
            <Row.Col span={6}>
              <div className='grid-col-box'>col3</div>
            </Row.Col>
            <Row.Col span={6}>
              <div className='grid-col-box'>col4</div>
            </Row.Col>
          </Row>
        </div>
        <div className='grid-row-box'>
          <Row gutter={10}>
            <Row.Col span={8}>
              <div className='grid-col-box'>col1</div>
            </Row.Col>
            <Row.Col span={0}>
              <div className='grid-col-box'>col2</div>
            </Row.Col>
            <Row.Col span={8}>
              <div className='grid-col-box'>col3</div>
            </Row.Col>
            <Row.Col span={8}>
              <div className='grid-col-box'>col4</div>
            </Row.Col>
          </Row>
        </div>
      </div>
    )
    expect(wrapper).toMatchSnapshot()
    // wrapper.unmount()
  })
  it('偏移布局', () => {
    const wrapper = render(
      <div>
        <div className='grid-row-box'>
          <Row gutter={10}>
            <Row.Col span={6}>
              <div className='grid-col-box'>col1</div>
            </Row.Col>
            <Row.Col span={6} offset={6}>
              <div className='grid-col-box'>col2</div>
            </Row.Col>
            <Row.Col span={6}>
              <div className='grid-col-box'>col3</div>
            </Row.Col>
          </Row>
        </div>
        <div className='grid-row-box'>
          <Row gutter={10}>
            <Row.Col span={6}>
              <div className='grid-col-box'>col1</div>
            </Row.Col>
            <Row.Col span={6} push={6}>
              <div className='grid-col-box'>col2</div>
            </Row.Col>
            <Row.Col span={6} pull={6}>
              <div className='grid-col-box'>col3</div>
            </Row.Col>
            <Row.Col span={6}>
              <div className='grid-col-box'>col4</div>
            </Row.Col>
          </Row>
        </div>
        <div className='grid-row-box'>
          <Row gutter={10}>
            <Row.Col span={6} offset={6}>
              <div className='grid-col-box'>col1</div>
            </Row.Col>
            <Row.Col span={6} pull={12}>
              <div className='grid-col-box'>col2</div>
            </Row.Col>
            <Row.Col span={6}>
              <div className='grid-col-box'>col4</div>
            </Row.Col>
          </Row>
        </div>
      </div>
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('响应式布局', () => {
    const wrapper = render(
      <div className='grid-row-box'>
        <Row gutter={{ xs: 8, sm: 10, md: 15, lg: 20, xl: 25, xxl: 30 }}>
          <Row.Col span={{ xs: 8, sm: 6, md: 4, lg: 6 }}>
            <div className='grid-col-box'>col1</div>
          </Row.Col>
          <Row.Col span={{ xs: 8, sm: 6, md: 5, lg: 10, xxl: 6 }}>
            <div className='grid-col-box'>col2</div>
          </Row.Col>
          <Row.Col span={{ xs: 8, sm: 6, md: 7, lg: 4, xl: 0, xxl: 6 }}>
            <div className='grid-col-box'>col3</div>
          </Row.Col>
          <Row.Col span={{ xs: 0, sm: 6, md: 8, lg: 4, xl: 8, xxl: 6 }}>
            <div className='grid-col-box'>col4</div>
          </Row.Col>
        </Row>
      </div>
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('响应式偏移', () => {
    const wrapper = render(
      <div>
        <div className='grid-row-box'>
          <Row gutter={10}>
            <Row.Col span={6}>
              <div className='grid-col-box'>col1</div>
            </Row.Col>
            <Row.Col span={6} offset={{ xs: 1, sm: 2, md: 3, lg: 4, xl: 5, xxl: 6 }}>
              <div className='grid-col-box'>col2</div>
            </Row.Col>
            <Row.Col span={6}>
              <div className='grid-col-box'>col3</div>
            </Row.Col>
          </Row>
        </div>
        <div className='grid-row-box'>
          <Row gutter={10}>
            <Row.Col span={6}>
              <div className='grid-col-box'>col1</div>
            </Row.Col>
            <Row.Col span={6}>
              <div className='grid-col-box'>col2</div>
            </Row.Col>
            <Row.Col span={6} push={{ xs: 1, sm: 2, md: 3, lg: 4, xl: 5, xxl: 6 }}>
              <div className='grid-col-box'>col3</div>
            </Row.Col>
          </Row>
        </div>
        <div className='grid-row-box'>
          <Row gutter={10}>
            <Row.Col span={6}>
              <div className='grid-col-box'>col1</div>
            </Row.Col>
            <Row.Col span={6}>
              <div className='grid-col-box'>col2</div>
            </Row.Col>
            <Row.Col span={6} pull={{ xs: 1, sm: 2, md: 3, lg: 4, xl: 5, xxl: 6 }} offset={6}>
              <div className='grid-col-box'>col3</div>
            </Row.Col>
          </Row>
        </div>
      </div>
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('flex布局', () => {
    const wrapper = render(
      <div>
        <div className='grid-row-box grid-row-box--flex'>
          <Row gutter={10} type={'flex'} align={'top'}>
            <Row.Col span={6}>
              <div className='grid-col-box'>col1</div>
            </Row.Col>
            <Row.Col span={6}>
              <div className='grid-col-box'>col2</div>
            </Row.Col>
            <Row.Col span={6}>
              <div className='grid-col-box'>col3</div>
            </Row.Col>
            <Row.Col span={6}>
              <div className='grid-col-box'>col4</div>
            </Row.Col>
          </Row>
        </div>
        <div className='grid-row-box grid-row-box--flex'>
          <Row gutter={10} type={'flex'} align={'middle'}>
            <Row.Col span={6}>
              <div className='grid-col-box'>col1</div>
            </Row.Col>
            <Row.Col span={6}>
              <div className='grid-col-box'>col2</div>
            </Row.Col>
            <Row.Col span={6}>
              <div className='grid-col-box'>col3</div>
            </Row.Col>
            <Row.Col span={6}>
              <div className='grid-col-box'>col4</div>
            </Row.Col>
          </Row>
        </div>
        <div className='grid-row-box grid-row-box--flex'>
          <Row gutter={10} type={'flex'} align={'bottom'}>
            <Row.Col span={6}>
              <div className='grid-col-box'>col1</div>
            </Row.Col>
            <Row.Col span={6}>
              <div className='grid-col-box'>col2</div>
            </Row.Col>
            <Row.Col span={6}>
              <div className='grid-col-box'>col3</div>
            </Row.Col>
            <Row.Col span={6}>
              <div className='grid-col-box'>col4</div>
            </Row.Col>
          </Row>
        </div>
        <div className='grid-row-box grid-row-box--flex'>
          <Row gutter={10} type={'flex'} justify={'start'}>
            <Row.Col span={6}>
              <div className='grid-col-box'>col1</div>
            </Row.Col>
            <Row.Col span={6}>
              <div className='grid-col-box'>col2</div>
            </Row.Col>
            <Row.Col span={6}>
              <div className='grid-col-box'>col3</div>
            </Row.Col>
          </Row>
        </div>
        <div className='grid-row-box grid-row-box--flex'>
          <Row gutter={10} type={'flex'} justify={'center'}>
            <Row.Col span={6}>
              <div className='grid-col-box'>col1</div>
            </Row.Col>
            <Row.Col span={6}>
              <div className='grid-col-box'>col2</div>
            </Row.Col>
            <Row.Col span={6}>
              <div className='grid-col-box'>col3</div>
            </Row.Col>
          </Row>
        </div>
        <div className='grid-row-box grid-row-box--flex'>
          <Row gutter={10} type={'flex'} justify={'end'}>
            <Row.Col span={6}>
              <div className='grid-col-box'>col1</div>
            </Row.Col>
            <Row.Col span={6}>
              <div className='grid-col-box'>col2</div>
            </Row.Col>
            <Row.Col span={6}>
              <div className='grid-col-box'>col3</div>
            </Row.Col>
          </Row>
        </div>
        <div className='grid-row-box grid-row-box--flex'>
          <Row gutter={10} type={'flex'}>
            <Row.Col span={6} order={3}>
              <div className='grid-col-box'>col1</div>
            </Row.Col>
            <Row.Col span={6} order={1}>
              <div className='grid-col-box'>col2</div>
            </Row.Col>
            <Row.Col span={6} order={2}>
              <div className='grid-col-box'>col3</div>
            </Row.Col>
          </Row>
        </div>
      </div>
    )
    expect(wrapper).toMatchSnapshot()
  })
})
