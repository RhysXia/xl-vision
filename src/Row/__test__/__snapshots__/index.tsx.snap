// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Row flex布局 1`] = `
<div>
  <div
    class="box"
  >
    <div
      class="xl-row-flex xl-row-flex--top"
      role="row"
      style="margin-left:-5px;margin-right:-5px"
    >
      <div
        class="xl-col xl-col-span-6"
        role="cell"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col1
        </div>
      </div>
      <div
        class="xl-col xl-col-span-6"
        role="cell"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col2
        </div>
      </div>
      <div
        class="xl-col xl-col-span-6"
        role="cell"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col3
        </div>
      </div>
      <div
        class="xl-col xl-col-span-6"
        role="cell"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col4
        </div>
      </div>
    </div>
  </div>
  <div
    class="box"
  >
    <div
      class="xl-row-flex xl-row-flex--middle"
      role="row"
      style="margin-left:-5px;margin-right:-5px"
    >
      <div
        class="xl-col xl-col-span-6"
        role="cell"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col1
        </div>
      </div>
      <div
        class="xl-col xl-col-span-6"
        role="cell"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col2
        </div>
      </div>
      <div
        class="xl-col xl-col-span-6"
        role="cell"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col3
        </div>
      </div>
      <div
        class="xl-col xl-col-span-6"
        role="cell"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col4
        </div>
      </div>
    </div>
  </div>
  <div
    class="box"
  >
    <div
      class="xl-row-flex xl-row-flex--bottom"
      role="row"
      style="margin-left:-5px;margin-right:-5px"
    >
      <div
        class="xl-col xl-col-span-6"
        role="cell"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col1
        </div>
      </div>
      <div
        class="xl-col xl-col-span-6"
        role="cell"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col2
        </div>
      </div>
      <div
        class="xl-col xl-col-span-6"
        role="cell"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col3
        </div>
      </div>
      <div
        class="xl-col xl-col-span-6"
        role="cell"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col4
        </div>
      </div>
    </div>
  </div>
  <div
    class="box"
  >
    <div
      class="xl-row-flex xl-row-flex--start"
      role="row"
      style="margin-left:-5px;margin-right:-5px"
    >
      <div
        class="xl-col xl-col-span-6"
        role="cell"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col1
        </div>
      </div>
      <div
        class="xl-col xl-col-span-6"
        role="cell"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col2
        </div>
      </div>
      <div
        class="xl-col xl-col-span-6"
        role="cell"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col3
        </div>
      </div>
    </div>
  </div>
  <div
    class="box"
  >
    <div
      class="xl-row-flex xl-row-flex--center"
      role="row"
      style="margin-left:-5px;margin-right:-5px"
    >
      <div
        class="xl-col xl-col-span-6"
        role="cell"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col1
        </div>
      </div>
      <div
        class="xl-col xl-col-span-6"
        role="cell"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col2
        </div>
      </div>
      <div
        class="xl-col xl-col-span-6"
        role="cell"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col3
        </div>
      </div>
    </div>
  </div>
  <div
    class="box"
  >
    <div
      class="xl-row-flex xl-row-flex--end"
      role="row"
      style="margin-left:-5px;margin-right:-5px"
    >
      <div
        class="xl-col xl-col-span-6"
        role="cell"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col1
        </div>
      </div>
      <div
        class="xl-col xl-col-span-6"
        role="cell"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col2
        </div>
      </div>
      <div
        class="xl-col xl-col-span-6"
        role="cell"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col3
        </div>
      </div>
    </div>
  </div>
  <div
    class="box"
  >
    <div
      class="xl-row-flex"
      role="row"
      style="margin-left:-5px;margin-right:-5px"
    >
      <div
        class="xl-col xl-col-order-3 xl-col-span-6"
        role="cell"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col1
        </div>
      </div>
      <div
        class="xl-col xl-col-order-1 xl-col-span-6"
        role="cell"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col2
        </div>
      </div>
      <div
        class="xl-col xl-col-order-2 xl-col-span-6"
        role="cell"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col3
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Row 偏移布局 1`] = `
<div>
  <div
    class="box"
  >
    <div
      class="xl-row"
      role="row"
      style="margin-left:-5px;margin-right:-5px"
    >
      <div
        class="xl-col xl-col-span-6"
        role="cell"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col1
        </div>
      </div>
      <div
        class="xl-col xl-col-offset-6 xl-col-span-6"
        role="cell"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col2
        </div>
      </div>
      <div
        class="xl-col xl-col-span-6"
        role="cell"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col3
        </div>
      </div>
    </div>
  </div>
  <div
    class="box"
  >
    <div
      class="xl-row"
      role="row"
      style="margin-left:-5px;margin-right:-5px"
    >
      <div
        class="xl-col xl-col-span-6"
        role="cell"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col1
        </div>
      </div>
      <div
        class="xl-col xl-col-push-6 xl-col-span-6"
        role="cell"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col2
        </div>
      </div>
      <div
        class="xl-col xl-col-pull-6 xl-col-span-6"
        role="cell"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col3
        </div>
      </div>
      <div
        class="xl-col xl-col-span-6"
        role="cell"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col4
        </div>
      </div>
    </div>
  </div>
  <div
    class="box"
  >
    <div
      class="xl-row"
      role="row"
      style="margin-left:-5px;margin-right:-5px"
    >
      <div
        class="xl-col xl-col-offset-6 xl-col-span-6"
        role="cell"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col1
        </div>
      </div>
      <div
        class="xl-col xl-col-pull-12 xl-col-span-6"
        role="cell"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col2
        </div>
      </div>
      <div
        class="xl-col xl-col-span-6"
        role="cell"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col4
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Row 响应式偏移 1`] = `
<div>
  <div
    class="box"
  >
    <div
      class="xl-row"
      role="row"
      style="margin-left:-5px;margin-right:-5px"
    >
      <div
        class="xl-col xl-col-span-6"
        role="cell"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col1
        </div>
      </div>
      <div
        class="xl-col xl-col-span-6"
        role="cell"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col2
        </div>
      </div>
      <div
        class="xl-col xl-col-span-6"
        role="cell"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col3
        </div>
      </div>
    </div>
  </div>
  <div
    class="box"
  >
    <div
      class="xl-row"
      role="row"
      style="margin-left:-5px;margin-right:-5px"
    >
      <div
        class="xl-col xl-col-span-6"
        role="cell"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col1
        </div>
      </div>
      <div
        class="xl-col xl-col-span-6"
        role="cell"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col2
        </div>
      </div>
      <div
        class="xl-col xl-col-span-6"
        role="cell"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col3
        </div>
      </div>
    </div>
  </div>
  <div
    class="box"
  >
    <div
      class="xl-row"
      role="row"
      style="margin-left:-5px;margin-right:-5px"
    >
      <div
        class="xl-col xl-col-span-6"
        role="cell"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col1
        </div>
      </div>
      <div
        class="xl-col xl-col-span-6"
        role="cell"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col2
        </div>
      </div>
      <div
        class="xl-col xl-col-offset-6 xl-col-span-6"
        role="cell"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col3
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Row 响应式布局 1`] = `
<div
  class="box"
>
  <div
    class="xl-row"
    role="row"
  >
    <div
      class="xl-col"
      role="cell"
    >
      <div
        class="col"
      >
        col1
      </div>
    </div>
    <div
      class="xl-col"
      role="cell"
    >
      <div
        class="col"
      >
        col2
      </div>
    </div>
    <div
      class="xl-col"
      role="cell"
    >
      <div
        class="col"
      >
        col3
      </div>
    </div>
    <div
      class="xl-col"
      role="cell"
    >
      <div
        class="col"
      >
        col4
      </div>
    </div>
  </div>
</div>
`;

exports[`Row 基本用法 1`] = `
<div>
  <div
    class="box"
  >
    <div
      class="xl-row"
      role="row"
      style="margin-left:-5px;margin-right:-5px"
    >
      <div
        class="xl-col xl-col-span-4"
        role="cell"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col1
        </div>
      </div>
      <div
        class="xl-col xl-col-span-5"
        role="cell"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col2
        </div>
      </div>
      <div
        class="xl-col xl-col-span-7"
        role="cell"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col3
        </div>
      </div>
      <div
        class="xl-col xl-col-span-8"
        role="cell"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col4
        </div>
      </div>
    </div>
  </div>
  <div
    class="box"
  >
    <div
      class="xl-row"
      role="row"
      style="margin-left:-5px;margin-right:-5px"
    >
      <div
        class="xl-col xl-col-span-6"
        role="cell"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col1
        </div>
      </div>
      <div
        class="xl-col xl-col-span-6"
        role="cell"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col2
        </div>
      </div>
      <div
        class="xl-col xl-col-span-6"
        role="cell"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col3
        </div>
      </div>
      <div
        class="xl-col xl-col-span-6"
        role="cell"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col4
        </div>
      </div>
    </div>
  </div>
  <div
    class="box"
  >
    <div
      class="xl-row"
      role="row"
      style="margin-left:-5px;margin-right:-5px"
    >
      <div
        class="xl-col xl-col-span-8"
        role="cell"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col1
        </div>
      </div>
      <div
        class="xl-col xl-col-span-0"
        role="cell"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col2
        </div>
      </div>
      <div
        class="xl-col xl-col-span-8"
        role="cell"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col3
        </div>
      </div>
      <div
        class="xl-col xl-col-span-8"
        role="cell"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col4
        </div>
      </div>
    </div>
  </div>
</div>
`;
